src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs

```diff
@@ -1,22 +1,98 @@
-﻿using System.Threading.Tasks;
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Abstractions;
 using Raven.Server.Routing;
 using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
 
 namespace Raven.Server.Documents.Handlers
 {
     public class DocumentReplicationRequestHandler : DatabaseRequestHandler
     {
-        [RavenAction("/databases/*/documentReplication", "GET", "/databases/{databaseName:string}/documentReplication")]
+        //an endpoint to establish replication websocket
+        [RavenAction("/databases/*/documentReplication", "GET",
+            @"/databases/{databaseName:string}/documentReplication?
+                srcDbId={databaseUniqueId:string}
+                &srcUrl={url:string}
+                &srcDbName={databaseName:string}
+                &lastSentEtag={etag:long}")]
         public async Task DocumentReplicationConnection()
         {
-            DocumentsOperationContext context;
-            
-            using (var webSocket = await HttpContext.WebSockets.AcceptWebSocketAsync())
-            using (ContextPool.AllocateOperationContext(out context))
+            var dbId = Guid.Parse(GetQueryStringValueAndAssertIfSingleAndNotEmpty("srcDbId")[0]);
+            var srcUrl = GetQueryStringValueAndAssertIfSingleAndNotEmpty("srcUrl")[0];
+            var srcDbName = GetQueryStringValueAndAssertIfSingleAndNotEmpty("srcDbName")[0];
+            long lastSentEtag;
+            if (!long.TryParse(GetQueryStringValueAndAssertIfSingleAndNotEmpty("lastSentEtag")[0], out lastSentEtag))
             {
-                //TODO : add disconnect at heartbeat timeout -> stop "while" loop at timeout
-                while (!Database.DatabaseShutdown.IsCancellationRequested)
-                {					
+                throw new ArgumentException("lastSentEtag should be a Int64 number, failed to parse...");
+            }
+
+            bool hasOtherSideClosedConnection = false;
+            bool shouldConnectBack;
+            var replicationExecuter = Database.DocumentReplicationLoader.RegisterNewConnectionFrom(dbId,
+                srcUrl,
+                srcDbName,
+                lastSentEtag,
+                out shouldConnectBack);
+            string ReplicationReceiveDebugTag = $"document-replication/receive <{replicationExecuter.Name}>";
+
+            using (var timeoutEvent = new ManualResetEventSlim())
+            {
+                var timeoutTimeSpan = Database.Configuration.Core.DatabaseOperationTimeout.AsTimeSpan;
+                using (var timeoutTimer = new Timer(_ => timeoutEvent.Set(), null, timeoutTimeSpan, Timeout.InfiniteTimeSpan))
```

Comment:
Why are you doing this? If we want to time out, we can use cancellation token, instead of this awkward arrangement.
And in general, I don't think that we do want this. We can just handle this via the TCP timeout already builtin.